text/0093-rocksdb-per-region.md

```diff
@@ -0,0 +1,196 @@
+
+## Detailed design
+
+Let's call an isolated rocksdb tablet. Now data is written into two storage raft engine and kv rocksdb. Tablets are used to replace kv rocksdb.
+
+### Data layout
+
+A directory “tablets” will be created inside the data directory to store the data of the tablet. The tablet will be named as ${region_id}_${tablet_index}. tablet_index is the raft log index that the tablet is corresponding to when being created. For example, when region 2 is created by split, its index is initialized to 5, so its name is 2_5. A replica of region 2 is catching up data by snapshot, and the snapshot index is 11, then after applying the snapshot, the tablet name becomes 2_11. Adding tablet_index is to allow quickly cleaning and applying of data. For example, there are still queries running on 2_5 while follower is about to apply a snapshot at index 11, the two directories can be kept and 2_5 can be deleted after all existing queries are finished.
+
+The old data directory “kv” will be deleted and not written any more.
+
+Tablets share env, blockcache, statics, rate limiters and compaction filters, so flow control, compaction concurrency and memory control are unified.
+
+```mermaid
+flowchart TB
+    subgraph Past
+    raft[(raft engine)]
+    kv[(KV DB)]
+    end
+    subgraph Dynamic regions
+    raft2[(raft engine)]
+    subgraph Tables
+    table0[(2_5)]
+    table1[(6_11)]
+    table2[(...)]
+    end
+    end
+```
+
+### Memtable management
+
+Each tablet has its own memtable, which can run out of memory quickly, so we need to manage memtables carefully. There are two ways to manage memtables: 1. mix storage, all tablets will share the same memtable, and it will be split and saved to their owned directory during flush. 2. separate storage, and flush should be triggered in time to reach a memory threshold. The design choose latter for:
+- It’s complicated to develop a mix storage, split in flush can produce many small files.
+- Separate storage can isolate writes for foreground writes and
+   - Because of isolation, there is no synchronization cost, so it’s faster in theory;
+   - If it’s written by a single thread, RocksDB can utilize insert hint to reduce the cost of comparison.
+
+Because bloom filters are allocated immediately by a factor of capacity when creating a memtable, if there are a lot of tablets, the memory consumption can be large even without any workload. In fact, configuration of v6.1 is set for mix storage, for separate storage, we can choose a smaller memtable size as it only contains data from one region, and the factor can also be reduced. We change it to 0.05 from 0.1 in the prototype.
+
+By separating RocksDB instances, there are only apply state and region state in the raft cf of KV DB. Writing two states every time is inefficient and triggers unnecessary compaction. We will move these stats to raft engine after disabling KV WAL.
```

Comment:
It's covered in the https://github.com/tikv/rfcs/pull/94.

Reply:
I think it's stated clearly in the "Disable KV WAL" section:
> Disabling KV WAL is a complicated design, it will be proposed as a separate RFC.

I will add a link when the #94 is merged.